beam/core/src/main/java/cz/o2/proxima/beam/core/BeamDataOperator.java

```diff
@@ -471,6 +481,23 @@ public boolean hasDirect() {
       // when limit is applied we must create a new source for each input
       ret = factory.apply(desc);
     }
+    if (!typesRegisteredFor(ret.getPipeline())) {
+      registerTypesFor(ret.getPipeline());
+    }
     return ret;
   }
+
+  private void registerTypesFor(Pipeline pipeline) {
+    CoderRegistry registry = pipeline.getCoderRegistry();
+    registry.registerCoderForClass(GlobalWindow.class, GlobalWindow.Coder.INSTANCE);
+    registry.registerCoderForClass(IntervalWindow.class, IntervalWindowCoder.of());
+    registry.registerCoderForClass(StreamElement.class, StreamElementCoder.of(repo));
+    registry.registerCoderForClass(EntityDescriptor.class, EntityDescriptorCoder.of(repo));
+    registry.registerCoderForClass(AttributeDescriptor.class, AttributeDescriptorCoder.of(repo));
+    typesRegistered.put(pipeline, true);
+  }
+
+  private boolean typesRegisteredFor(Pipeline pipeline) {
+    return Optional.ofNullable(typesRegistered.get(pipeline)).orElse(false);
```

Comment:
may you could just turn 'typesRegistered' into a `Set` an check for element presence over here? I'd say this is just kind of complicated presence check...

Reply:
emmm .... :)

Reply:
fixed